Classes/common/touchdb/TD_Database.m

```diff
@@ -438,24 +577,43 @@ - (BOOL)close
     _activeReplicators = nil;
 
     [_fmdbQueue close];
+    _fmdbQueue = nil;
+    _keyProviderToOpenDB = nil;
 
     _open = NO;
     _transactionLevel = 0;
+    
     return YES;
 }
 
-- (BOOL)deleteDatabase:(NSError**)outError
+- (BOOL)deleteDatabase:(NSError **)outError
 {
-    CDTLogInfo(CDTDATASTORE_LOG_CONTEXT, @"Deleting %@", _path);
     [[NSNotificationCenter defaultCenter] postNotificationName:TD_DatabaseWillBeDeletedNotification
                                                         object:self];
-    if (_open) {
-        if (![self close]) return NO;
-    } else if (!self.exists) {
-        return YES;
+
+    if (_open && ![self close]) {
+        CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"Database at path %@ could not be closed", _path);
+
+        return NO;
+    }
+
+    return [[self class] deleteClosedDatabaseAtPath:_path error:outError];
+}
+
++ (BOOL)deleteClosedDatabaseAtPath:(NSString *)path error:(NSError **)outError
+{
+    CDTLogInfo(CDTDATASTORE_LOG_CONTEXT, @"Deleting %@", path);
+    
+    BOOL success = YES;
+
+    if ([TD_Database existsDatabaseAtPath:path]) {
+        NSString *attachmentsPath = [TD_Database attachmentStorePathWithDatabasePath:path];
+
+        success =
+            (removeItemIfExists(path, outError) && removeItemIfExists(attachmentsPath, outError));
```

Comment:
A crash after `removeItemIfExists(path, outError)` would leave the attachments data hanging around and one would never be able to delete it through this method because the conditional would fail because the database would be gone.